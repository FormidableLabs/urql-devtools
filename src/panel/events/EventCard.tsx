import React, { FC, useContext, useCallback } from "react";
import styled, { ThemeContext } from "styled-components";
import { OperationEvent } from "../../types";
import { OperationContext } from "../context";

/** Shows basic information about an operation. */
export const EventCard: FC<{ operation: OperationEvent }> = ({ operation }) => {
  const theme = useContext(ThemeContext);
  const {
    selectedOperation,
    selectOperation,
    clearSelectedOperation
  } = useContext(OperationContext);

  const handleContainerClick = useCallback(() => {
    selectedOperation === operation
      ? clearSelectedOperation()
      : selectOperation(operation);
  }, [operation, selectedOperation, selectOperation]);

  const colors = {
    subscription: theme.orange,
    teardown: theme.grey,
    mutation: theme.purple,
    query: theme.lightBlue,
    response: theme.green,
    error: theme.red
  };

  const name =
    operation.type === "operation"
      ? operation.data.operationName
      : operation.type;
  const date = formatDate(operation.timestamp);
  const info =
    operation.type === "operation"
      ? operation.data.context.devtools.source
      : operation.data.operation.context.devtools.source;
  const key =
    operation.type === "operation"
      ? operation.data.key
      : operation.data.operation.key;

  return (
    <Container onClick={handleContainerClick}>
      <Indicator style={{ backgroundColor: colors[name] }} />
      <OperationName>{capitalize(name)}</OperationName>
      <OperationTime>{date}</OperationTime>
      <OperationAddInfo>{info || "Unknown"}</OperationAddInfo>
      <OperationKey>{key}</OperationKey>
    </Container>
  );
};

const formatDate = (date: number) => {
  const d = new Date(date);
  const padded = (n: number) => String(n).padStart(2, "0");

  return `${padded(d.getHours())}:${padded(d.getMinutes())}:${padded(
    d.getSeconds()
  )}`;
};

const capitalize = (s: string) => `${s.charAt(0).toUpperCase()}${s.slice(1)}`;

const OperationName = styled.h3`
  color: rgba(255, 255, 255, 0.9);
  font-size: 15px;
  margin: 0;
  width: 50%;

  @media (max-width: ${props => props.theme.breakpoints.sm.max}) {
    margin-bottom: 10px;
  }

  @media (min-width: ${props => props.theme.breakpoints.md.min}) {
    color: rgba(255, 255, 255, 0.8);
    order: 1;
    font-size: 13px;
    width: auto;
    flex-basis: 4;
    width: 20%;
  }
`;

const OperationTime = styled.p`
  color: rgba(255, 255, 255, 0.8);
  font-size: 14px;
  margin: 0;
  width: 50%;
  text-align: right;

  @media (min-width: ${props => props.theme.breakpoints.md.min}) {
    order: 4;
    font-size: 13px;
    width: 20%;
  }
`;

const OperationAddInfo = styled.p`
  color: rgba(255, 255, 255, 0.7);
  margin: 0;
  width: 50%;

  @media (min-width: ${props => props.theme.breakpoints.md.min}) {
    color: rgba(255, 255, 255, 0.8);
    order: 2;
    font-size: 13px;
    width: 25%;
  }
`;

const OperationKey = styled.p`
  margin: 0;

  @media (max-width: ${props => props.theme.breakpoints.sm.max}) {
    color: rgba(255, 255, 255, 0.7);
    width: 50%;
    text-align: right;
  }

  @media (min-width: ${props => props.theme.breakpoints.md.min}) {
    color: rgba(255, 255, 255, 0.8);
    order: 3;
    font-size: 13px;
    width: 25%;
  }
`;

const Indicator = styled.div`
  position: absolute;
  left: 0;
  width: 5px;
  top: 0;
  bottom: 0;
`;

const Container = styled.div`
  position: relative;
  background-color: ${props => props.theme.cardBg};
  width: auto;
  height: auto;
  display: flex;
  flex-direction: row;
  padding: 10px 15px;
  margin: 10px 0;

  @media (max-width: ${props => props.theme.breakpoints.sm.max}) {
    flex-wrap: wrap;
    align-items: baseline;
  }

  @media (min-width: ${props => props.theme.breakpoints.md.min}) {
    align-items: center;
    justify-content: space-between;
  }
`;
