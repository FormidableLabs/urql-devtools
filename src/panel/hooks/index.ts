export * from "./useThemeContext";
export * from "./useOrientationWatcher";
