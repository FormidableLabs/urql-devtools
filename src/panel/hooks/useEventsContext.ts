import { useContext } from "react";
import { EventsContext } from "../context";

export const useEventsContext = () => useContext(EventsContext);
