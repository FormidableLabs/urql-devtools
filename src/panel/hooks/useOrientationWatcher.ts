import { useEffect, useState, useMemo } from "react";

const aspectQuery = window.matchMedia("(min-aspect-ratio: 1/1)");

interface OrientationWatcher {
  isPortrait: boolean;
  isLandscape: boolean;
}

export const useOrientationWatcher = (): OrientationWatcher => {
  const [state, setState] = useState(aspectQuery.matches);

  useEffect(() => {
    const listener = () => setState(aspectQuery.matches);
    aspectQuery.addEventListener("change", listener);
    return () => aspectQuery.removeEventListener("change", listener);
  }, []);

  return useMemo(() => ({ isPortrait: !state, isLandscape: state }), [state]);
};
