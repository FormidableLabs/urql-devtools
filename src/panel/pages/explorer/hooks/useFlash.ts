import { useSpring } from "react-spring";
import { useCallback } from "react";

const defaultState = { background: "rgba(255, 255, 255, 0)" };
const flashState = { background: "rgba(255, 255, 255, 1)" };

type Flash = () => void;

type UseFlashResponse = [any, Flash];

/** Hook for flashing a screen element */
export const useFlash = (): UseFlashResponse => {
  const [props, setSpring] = useSpring(() => ({
    config: { duration: 300 },
  }));

  const flash = useCallback(
    () =>
      setSpring({
        from: defaultState,
        to: [flashState, defaultState],
      } as any),
    [setSpring]
  );

  return [props, flash];
};
