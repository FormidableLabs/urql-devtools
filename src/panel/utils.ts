import { DevtoolsExchangeOutgoingMessage } from "@urql/devtools";

const filterEventTypes = <T extends DevtoolsExchangeOutgoingMessage["type"]>(
  types: T[]
) => {};

type FilterType<T, A> = A extends { type: T } ? true : false;

const excludeEventTypes = <T extends DevtoolsExchangeOutgoingMessage["type"]>(
  types: T[]
) => <E extends DevtoolsExchangeOutgoingMessage>(
  event: E
): FilterType<T, E> => {
  if (types.indexOf(event.type as any)) {
    return true;
  }

  return false;
};
