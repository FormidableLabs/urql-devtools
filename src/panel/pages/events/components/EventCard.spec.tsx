jest.mock("../../hooks", () => ({
  useThemeContext: jest.fn(),
  useEventsContext: jest.fn()
}));
jest.mock("react-spring");
import * as React from "react";
import { shallow } from "enzyme";
import { mocked } from "ts-jest/utils";
import { useThemeContext, useEventsContext } from "../../../hooks";
import { EventCard } from "./EventCard";

const theme = new Proxy({}, { get: (t, p) => [p] });
const events = {
  selectedEvent: undefined,
  selectEvent: jest.fn(),
  clearSelectedEvent: jest.fn(),
  addFilter: jest.fn()
};

mocked(useThemeContext).mockReturnValue(theme as any);
mocked(useEventsContext).mockReturnValue(events as any);

const event = {
  type: "mutation",
  key: 1234,
  source: "ComponentSource",
  timestamp: 123123,
  panels: [
    { name: "query", data: "data" },
    { name: "variables", data: {} },
    { name: "response", data: {} },
    { name: "meta", data: {} }
  ]
} as const;

describe("on mount", () => {
  describe("active event", () => {
    it("matches snapshot", () => {
      const wrapper = shallow(
        <EventCard event={event} active={true} canFilter={false} />
      );
      expect(wrapper).toMatchSnapshot();
    });
  });

  describe("inactive event", () => {
    it("matches snapshot", () => {
      const wrapper = shallow(
        <EventCard event={event} active={false} canFilter={false} />
      );
      expect(wrapper).toMatchSnapshot();
    });
  });
});

describe("on event click", () => {
  it("calls selectEvent", () => {
    const wrapper = shallow(
      <EventCard event={event} active={false} canFilter={false} />
    );
    wrapper.find("Container").simulate("click", {});
    expect(events.selectEvent).toBeCalledTimes(1);
    expect(events.selectEvent).toBeCalledWith(event);
  });
});
