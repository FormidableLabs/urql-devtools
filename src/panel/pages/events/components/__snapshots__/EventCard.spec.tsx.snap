// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`on mount active event matches snapshot 1`] = `
<Container
  aria-selected={true}
  onClick={[Function]}
>
  <Indicator
    style={
      Object {
        "backgroundColor": "purple",
      }
    }
  />
  <OperationName
    isActive={false}
    onClick={[Function]}
  >
    mutation
  </OperationName>
  <OperationTime>
    01:02:03
  </OperationTime>
  <OperationAddInfo
    isActive={false}
    onClick={[Function]}
  >
    ComponentSource
  </OperationAddInfo>
  <OperationKey
    isActive={false}
    onClick={[Function]}
  >
    1234
  </OperationKey>
</Container>
`;

exports[`on mount inactive event matches snapshot 1`] = `
<Container
  aria-selected={false}
  onClick={[Function]}
>
  <Indicator
    style={
      Object {
        "backgroundColor": "purple",
      }
    }
  />
  <OperationName
    isActive={false}
    onClick={[Function]}
  >
    mutation
  </OperationName>
  <OperationTime>
    01:02:03
  </OperationTime>
  <OperationAddInfo
    isActive={false}
    onClick={[Function]}
  >
    ComponentSource
  </OperationAddInfo>
  <OperationKey
    isActive={false}
    onClick={[Function]}
  >
    1234
  </OperationKey>
</Container>
`;
